### PR TITLE
README: Add user documentation and badges

### LINKED ISSUE #1: User-level documentation
We need documentation for users, using some automated generation tool, eg. sphinx.

### PR DESCRIPTION
**Pull request by mschmidt87**
This PR adds user-level documentation to the README:
- A simple guide of how to set up a CGP experiment
- A brief description of the implemented data structures

+ some Python badges.

Best way to test this is to look at the rendered README: https://github.com/jakobj/python-gp/tree/doc/README_user_doc
This fixes #1 .

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,7 +1,117 @@
...
+4. Define a callback function to record information about the evolution of the population:
+```
+def record_history(pop, history):
+    keys = ['fitness']
+    for key in keys:
+        if key not in history:
+            history[key] = []
+    history['fitness'].append(pop.fitness_parents())
+```
+5. Use the `evolve` function that ties everything together and executes the evolution:
```
Comment (jakobj) [inline README.md]:
```suggestion
5. Use the `evolve` function from the high-level API that ties everything together and executes the evolution:
```

File: README.md [hunk 2]
```diff
@@ -1,7 +1,117 @@
...
+```
+history = gp.evolve(pop, obj, ea, params['max_generations'],
+                       params['min_fitness'], record_history=record_history, print_progress=True)
+```
+
```
Comment (jakobj) [inline README.md]: do we need so many line breaks? ;)

File: README.md [hunk 3]
```diff
@@ -1,7 +1,117 @@
...
+
+
+Data structures
+===============
+
+The Cartesian genetic programming framework is implemented with the following data structures:
```
Comment (jakobj) [inline README.md]:
```suggestion
The Cartesian Genetic Programming framework is implemented with the following data structures:
```

File: README.md [hunk 4]
```diff
@@ -1,7 +1,117 @@
...
+
+- `CGPPopulation` implements a population of `CGPIndividual` parent individuals and defines the standard procedures to create offspring from parent individuals, mutation and crossover.
```
Comment (jakobj) [inline README.md]:
```suggestion
- `CGPPopulation` represents a population of `CGPIndividual` individuals and defines the standard procedures to create offspring from these as well as mutation and crossover.
```

File: README.md [hunk 5]
```diff
@@ -1,7 +1,117 @@
...
+	- `to_torch` creates a torch.nn.Module child class.
+	- `to_func` creates a Python callable.
+- Primitives are operations that are combined in a computational graphs to produce an output value from given input values. Any operation is supported, the library currently implements:
+   - Addition: `CGPAdd`
+   - Subtraction: `CGPSub`
+   - Multiplication: `CGPMul`
+   - Division: `CGPDiv`
+   - Exponentiation: `CGPPow`
+   - Constants: `CGPConstantFloat`
+   These operations are implemented as child classes of `CGPNode` and are grouped by the umbrella class `CGPPrimitives` inside the `CGPGenome`.
```
Comment (jakobj) [inline README.md]: do we need to mention `CGPPrimitives`?

File: README.md [hunk 6]
```diff
@@ -1,7 +1,117 @@
...
+```
+5. Use the `evolve` function that ties everything together and executes the evolution:
+```
+history = gp.evolve(pop, obj, ea, params['max_generations'],
+                       params['min_fitness'], record_history=record_history, print_progress=True)
+```
+
+
+
+Data structures
```
Comment (jakobj) [inline README.md]: do we need such a detailed discussion of the relevant classes here?

File: README.md [hunk 7]
```diff
@@ -1,7 +1,117 @@
...
+    -------
+    gp.CGPIndividual
+        Modified individual with updated fitness value.
+    """
+	# Compute the fitness value
+	fitness = ...
+	individual.fitness = fitness
+	return individual
+```
+2. Define parameters for the genome, the population and the evolutionary algorithm
```
Comment (jakobj) [inline README.md]: either all headings should be bold or none

File: README.md [hunk 8]
```diff
@@ -1,7 +1,117 @@
...
+===========
+
+A complete example of CGP applied to a regression problem can be found in `examples/example_evo_regression.py`.
+
+The basic steps to be taken are:
+
+1. **Define an objective function**. 
+
+   The objective needs to take an individual as input variable and update the `fitness` member of the individual.
+```
```
Comment (jakobj) [inline README.md]: is it possible to render this as python code?

File: README.md [hunk 9]
```diff
@@ -1,7 +1,117 @@
...
+[![Python3.6](https://img.shields.io/badge/python-3.6-red.svg)](https://www.python.org/downloads/release/python-369/)
+[![Python3.7](https://img.shields.io/badge/python-3.7-red.svg)](https://www.python.org/)
+[![Python3.8](https://img.shields.io/badge/python-3.8-red.svg)](https://www.python.org/)
+[![GPL license](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-3.0.html)
+
+Cartesian genetic programming (CGP) in Python.
+
+This library implements Cartesian genetic programming in pure Python. It implements Python data structures to represent computational graphs in the Cartesian Genetic programming and provides one evolutionary algorithm,  (mu + lambda) evolution strategies, to evolve a population of computational graphs given an objective function.
+
+![CGP Sketch](cgp-sketch.png)
```
[the remark this answers is not in the document]
Comment (mschmidt87) [reply README.md]: I increased the resolution and resized the image: 14357ae496a338d747d77abdc0709d467286117c

File: README.md (hunk 5 above)
Comment (mschmidt87) [reply README.md]: I would like to mention it because e.g. even for me, it wasn't immediately clear what this class is used for (see #16 )

File: README.md (hunk 6 above)
Comment (mschmidt87) [reply README.md]: Why not? I think it is useful for someone who likes to understand this library.

File: README.md (hunk 7 above)
Comment (mschmidt87) [reply README.md]: I unbolded the first heading.

File: README.md (hunk 8 above)
Comment (mschmidt87) [reply README.md]: It is! Awesome! 5e5d4ff94796ad7a98e2baf5b5a8b0d5cc3508e7

**Comment by mschmidt87** [comment]
Committed all suggestions and addressed comments.

File: README.md [hunk 10]
```diff
@@ -1,7 +1,116 @@
 python-gp
 ===========
+[![Python3.6](https://img.shields.io/badge/python-3.6-red.svg)](https://www.python.org/downloads/release/python-369/)
+[![Python3.7](https://img.shields.io/badge/python-3.7-red.svg)](https://www.python.org/)
+[![Python3.8](https://img.shields.io/badge/python-3.8-red.svg)](https://www.python.org/)
+[![GPL license](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/old-licenses/gpl-3.0.html)
+
+Cartesian Genetic Programming (CGP) in Python.
+
+This library implements Cartesian Genetic Programming in pure Python. It provides Python data structures to represent and evolve a two-dimensional directed graph (the genotype) that can be translated into a computational graph (the phenotype) implementing a particular mathematical expression. The current implementation uses an evolutionary algorithm,  (mu + lambda) evolution strategies, to evolve a population of Cartesian graphs in order to optimize an objective function.
```
Comment (jakobj) [inline README.md]:
```suggestion
This library implements Cartesian Genetic Programming for symbolic regression in pure Python. It provides Python data structures to represent and evolve a two-dimensional directed graph (the genotype) that can be translated into a computational graph (the phenotype) implementing a particular mathematical expression. The current implementation uses an evolutionary algorithm,  specifically (mu + lambda) evolution strategies, to evolve a population of symbolic expressions in order to optimize an objective function.

```

File: README.md [hunk 11]
```diff
@@ -1,7 +1,116 @@
...
+
+      Returns
+      -------
+      gp.CGPIndividual
+          Modified individual with updated fitness value.
+      """
+      # Compute the fitness value
+      fitness = ...
+	  individual.fitness = fitness
+	  return individual
```
Comment (jakobj) [inline README.md]:
```suggestion
	  individual.fitness = ...
	  return individual
```

File: README.md (hunk 5 above)
Comment (jakobj) [reply README.md]: i think as a user you shouldn't even see this class, right?

File: README.md (hunk 6 above)
Comment (jakobj) [reply README.md]: my main concern is maintenance: as the library develops this might become outdated and we need to keep track of it manually. for such "low level" documentation maybe something automated like sphinx is the better option?

File: README.md (hunk 11 above)
Comment (jakobj) [reply README.md]: i'm not sure whether you rejected this or it just disappeared?

File: README.md [hunk 12]
```diff
@@ -1,7 +1,116 @@
...
+```
+2. Define parameters for the genome, the population and the evolutionary algorithm
+```python
+params = {
+     'seed': 8188212,
+     'n_threads': 1,
```
Comment (jakobj) [inline README.md]: change to `n_processes` and add to `ea_params`

File: README.md (hunk 10 above)
Comment (jakobj) [reply README.md]: in the manuscript we refer to the flattened 1d array of integers as the genotype, here we refer to the 2d directed graph as the genotype. i think we should settle on one; i'd prefer the latter and adapt the manuscript accordingly. the flattened array of ints is an implementation detail imho

File: README.md (hunk 11 above)
Comment (mschmidt87) [reply README.md]: Your original suggestion looked different, that's why I rejected. This is fine.

File: README.md (hunk 12 above)
Comment (mschmidt87) [reply README.md]: Done. cef7b7a23477554b7cd4dfbba200927ed5fb897a

**Comment by mschmidt87** [comment]
I addressed the comments and removed the data structures section.

File: README.md (hunk 10 above)
Comment (mschmidt87) [reply README.md]: Done.

File: README.md [hunk 13]
```diff
@@ -1,7 +1,93 @@
...
+Basic usage
+===========
+
+A simple example of CGP applied to a symbolic regression problem can be found in `examples/example_evo_regression.py`.
+
+The basic steps are:
+
+1. Define an objective function. 
+
+   The objective function needs to take an individual as input variable and update the `fitness` of the individual.
```
Comment (jakobj) [inline README.md]:
```suggestion
   The objective function takes an individual as an argument and updates the `fitness` of the individual.
```

File: README.md [hunk 14]
```diff
@@ -1,7 +1,93 @@
...
+          Individual of the Cartesian Genetic Programming Framework.
+
+      Returns
+      -------
+      gp.CGPIndividual
+          Modified individual with updated fitness value.
+      """
+      # Compute the fitness value
+	  individual.fitness = ...
+	  return individual
```
Comment (jakobj) [inline README.md]:
```suggestion
	individual.fitness = ...
	return individual
```

File: README.md [hunk 15]
```diff
@@ -1,7 +1,93 @@
...
+     'population_params': {
+          'n_parents': 10,
+          'mutation_rate': 0.5,
+      },
+
+     'ea_params': {
+         'n_offsprings': 10,
+         'n_breeding': 10,
+         'tournament_size': 1,
+		 'n_processes': 2},
```
Comment (jakobj) [inline README.md]:
```suggestion
         'tournament_size': 1,
         'n_processes': 2},
```

File: README.md [hunk 16]
```diff
@@ -1,7 +1,93 @@
...
+ea = gp.ea.MuPlusLambda(**params['ea_params'])
+```
+4. Define a callback function to record information about the evolution of the population:
+```python
+def record_history(pop, history):
+        keys = ['fitness']
+        for key in keys:
+            if key not in history:
+                history[key] = []
+        history['fitness'].append(pop.fitness_parents())
```
Comment (jakobj) [inline README.md]:
```suggestion
history = {}
history["fitness_parents"] = []
def recording_callback(pop):
    history["fitness_parents"].append(pop.fitness_parents())
```

Comment (jakobj) [reply README.md]: this reflects the new history recording method

File: README.md [hunk 17]
```diff
@@ -1,7 +1,93 @@
...
+```
+5. Use the `evolve` function from the high-level API that ties everything together and executes the evolution:
+```python
+history = gp.evolve(pop, obj, ea, params['max_generations'],
+                       params['min_fitness'], record_history=record_history, print_progress=True)
```
Comment (jakobj) [inline README.md]:
```suggestion
                       params['min_fitness'], print_progress=True, callback=recording_callback)
```

**Comment by mschmidt87** [comment]
Added @jakobj 's suggestions.

**Comment by jakobj** [comment]
travis is coughing a lot, let's wait for them, then i'll merge
